### PR TITLE
mastercard encryption

### PR DESCRIPTION
**Pull request by mateusmrangel**
Hello everyone  
With this merge request, the lib will support MasterCard encryption. See the usage on the mastercard_encryption/mastercard_encryption_test.go file.  
I used as a base the Java implementation of the client-encryption library

### CODE REVIEW

**Comment by mateusmrangel** [comment]
@danny-gallagher

File: field_level_encryption/field_level_encryption_config_builder.go [hunk 1]
```diff
@@ -0,0 +1,121 @@
...
+	cb.decryptionKey = decryptionKey
+	return cb
+}
+
+func (cb *FieldLevelEncryptionConfigBuilder) WithEncryptionPath(jsonPathIn string, jsonPathOut string) *FieldLevelEncryptionConfigBuilder {
+	cb.encryptionPaths[jsonPathIn] = jsonPathOut
+	return cb
+}
+
+func (cb *FieldLevelEncryptionConfigBuilder) WithDencryptionPath(jsonPathIn string, jsonPathOut string) *FieldLevelEncryptionConfigBuilder {
```
Comment (danny-gallagher) [inline field_level_encryption/field_level_encryption_config_builder.go]: Decryption is mis-spelled here

"WithDencryptionPath" -> "WithDecryptionPath"

**Comment by danny-gallagher** [comment]
Hi @mateusmrangel,

Thanks for creating this PR! There are a couple of things that will need to be added before we can merge this PR.
- If you take a look in "http_client_interceptor.go", you can see that we provide an interceptor which can be used alongside an openapi generated client to intercept and encrypt a request before it enters transit. We'll need to add support in this interceptor for Field Level Encryption.
- The README will need to be updated to reflect the new Field Level Encryption functionality. (See the java library for how this should look)
- The tests for this new addition should test all of the components of the library that the java library does.

If you have any questions, please feel free to reach out!

File: field_level_encryption/field_level_encryption_config_builder.go (hunk 1 above)
Comment (mateusmrangel) [reply field_level_encryption/field_level_encryption_config_builder.go]: fixed in [3f2b1a7](https://github.com/Mastercard/client-encryption-go/pull/6/commits/3f2b1a7106e469cb1bbc24a59bd8ecb71ea6c4bc)

**Comment by mateusmrangel** [comment]
> Hi @mateusmrangel,
> 
> Thanks for creating this PR! There are a couple of things that will need to be added before we can merge this PR.
> 
> * If you take a look in "http_client_interceptor.go", you can see that we provide an interceptor which can be used alongside an openapi generated client to intercept and encrypt a request before it enters transit. We'll need to add support in this interceptor for Field Level Encryption.
> * The README will need to be updated to reflect the new Field Level Encryption functionality. (See the java library for how this should look)
> * The tests for this new addition should test all of the components of the library that the java library does.
> 
> If you have any questions, please feel free to reach out!

Ok, I am going to work on it

File: mastercard_encryption/mastercard_encryption_test.go [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+		WithEncryptionKeyFingerprintFieldName("publicKeyFingerprint").
+		WithOaepPaddingDigestAlgorithmFieldName("oaepPaddingDigestAlgorithm").
+		WithFieldValueEncoding("HEX").
+		Build()
+	params := field_level_encryption.Generate(flConfig)
+
+	encryptedPayload := EncryptPayload(payload, flConfig, params)
+	fmt.Println(encryptedPayload)
+
+	decryptedPayload := DecryptPayload(encryptedPayload, flConfig, params)
```
Comment (danny-gallagher) [inline mastercard_encryption/mastercard_encryption_test.go:89]: The 'params' for decrypting the payload should be extracted from the actual request body/headers as they are in our other libraries.

File: field_level_encryption/field_level_encryption_params.go [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+	flep.OaepPaddingDigestAlgorithmValue = oaepPaddingDigestAlgorithmValue
+	flep.Config = config
+
+	return &flep
+}
+
+func Generate(config *FieldLevelEncryptionConfig) *FieldLevelEncryptionParams {
+	//// Generate a random IV
+	ivParameterSpec := aes_encryption.GenerateCEK(16 * 8)
+	ivSpecValue := utils.HexUrlEncode(ivParameterSpec)
```
Comment (danny-gallagher) [inline field_level_encryption/field_level_encryption_params.go:38]: In our other libraries (Java for example), we allow the user to specify the FieldValueEncoding through the config. In your implementation it's hard coded to always use Hex.

File: mastercard_encryption/mastercard_encryption.go [hunk 4]
```diff
@@ -0,0 +1,76 @@
...
+
+func DecryptPayload(encryptedPayload string, config *field_level_encryption.FieldLevelEncryptionConfig, params *field_level_encryption.FieldLevelEncryptionParams) string {
+	jsonPayload, _ := gabs.ParseJSON([]byte(encryptedPayload))
+	for jsonPathIn, jsonPathOut := range config.GetDecryptionPaths() {
+		jsonPayload = decryptPayloadPath(jsonPayload, jsonPathIn, jsonPathOut, params)
+	}
+	return jsonPayload.String()
+}
+
+func decryptPayloadPath(jsonPayload *gabs.Container, jsonPathIn string, jsonPathOut string, params *field_level_encryption.FieldLevelEncryptionParams) *gabs.Container {
```
Comment (danny-gallagher) [inline mastercard_encryption/mastercard_encryption.go:61]: The 'jsonPathOut' isn't used in this function. This property should be used to include the encrypted payload in a specific path in the payload (See the Java encryption library)

**Review by danny-gallagher — COMMENTED** [review]
In our other libraries, it's possible to include the encryption properties in the header of a request/response. You can see this functionality in the Java library for reference.

The functionality for this library should be identical to all of our other libraries.
You can include all the tests that are in our Java encryption library to ensure you're covering all possible cases.

**Comment by danny-gallagher** [comment]
Addressing this PR in the following PR: https://github.com/Mastercard/client-encryption-go/pull/14

**Pull request closed without merging**